// @react-navigation/bottom-tabs/lib/commonjs/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "createBottomTabNavigator", {
  enumerable: true,
  get: function get() {
    return _createBottomTabNavigator.default;
  }
});
Object.defineProperty(exports, "BottomTabView", {
  enumerable: true,
  get: function get() {
    return _BottomTabView.default;
  }
});
Object.defineProperty(exports, "BottomTabBar", {
  enumerable: true,
  get: function get() {
    return _BottomTabBar.default;
  }
});
Object.defineProperty(exports, "BottomTabNavigationOptions", {
  enumerable: true,
  get: function get() {
    return _types.BottomTabNavigationOptions;
  }
});
Object.defineProperty(exports, "BottomTabNavigationProp", {
  enumerable: true,
  get: function get() {
    return _types.BottomTabNavigationProp;
  }
});
Object.defineProperty(exports, "BottomTabBarProps", {
  enumerable: true,
  get: function get() {
    return _types.BottomTabBarProps;
  }
});
Object.defineProperty(exports, "BottomTabBarOptions", {
  enumerable: true,
  get: function get() {
    return _types.BottomTabBarOptions;
  }
});

var _createBottomTabNavigator = _interopRequireDefault(require("./navigators/createBottomTabNavigator"));

var _BottomTabView = _interopRequireDefault(require("./views/BottomTabView"));

var _BottomTabBar = _interopRequireDefault(require("./views/BottomTabBar"));

var _types = require("./types");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
