// @react-navigation/routers/lib/commonjs/CommonActions.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.goBack = goBack;
exports.navigate = navigate;
exports.reset = reset;
exports.setParams = setParams;

function goBack() {
  return {
    type: 'GO_BACK'
  };
}

function navigate() {
  if (typeof (arguments.length <= 0 ? undefined : arguments[0]) === 'string') {
    return {
      type: 'NAVIGATE',
      payload: {
        name: arguments.length <= 0 ? undefined : arguments[0],
        params: arguments.length <= 1 ? undefined : arguments[1]
      }
    };
  } else {
    const payload = arguments.length <= 0 ? undefined : arguments[0];

    if (!payload.hasOwnProperty('key') && !payload.hasOwnProperty('name')) {
      throw new Error('While calling navigate with an object as the argument, you need to specify name or key');
    }

    return {
      type: 'NAVIGATE',
      payload
    };
  }
}

function reset(state) {
  return {
    type: 'RESET',
    payload: state
  };
}

function setParams(params) {
  return {
    type: 'SET_PARAMS',
    payload: {
      params
    }
  };
}
