// @react-navigation/core/lib/module/useNavigationHelpers.js
function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { _defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

import * as React from 'react';
import { CommonActions } from '@react-navigation/routers';
import NavigationContext from './NavigationContext';
import { PrivateValueStore } from './types'; // This is to make TypeScript compiler happy
// eslint-disable-next-line babel/no-unused-expressions

PrivateValueStore;

/**
 * Navigation object with helper methods to be used by a navigator.
 * This object includes methods for common actions as well as methods the parent screen's navigation object.
 */
export default function useNavigationHelpers(_ref) {
  let {
    onAction,
    getState,
    emitter,
    router
  } = _ref;
  const parentNavigationHelpers = React.useContext(NavigationContext);
  return React.useMemo(() => {
    const dispatch = action => {
      const payload = typeof action === 'function' ? action(getState()) : action;
      const handled = onAction(payload);

      if (!handled && process.env.NODE_ENV !== 'production') {
        console.error("The action '".concat(payload.type, "' with payload '").concat(JSON.stringify(payload.payload), "' was not handled by any navigator. If you are trying to navigate to a screen, check if the screen exists in your navigator. If you're trying to navigate to a screen in a nested navigator, see https://reactnavigation.org/docs/en/nesting-navigators.html#navigating-to-a-screen-in-a-nested-navigator."));
      }
    };

    const actions = _objectSpread({}, router.actionCreators, {}, CommonActions);

    const helpers = Object.keys(actions).reduce((acc, name) => {
      // @ts-ignore
      acc[name] = function () {
        return dispatch(actions[name](...arguments));
      };

      return acc;
    }, {});
    return _objectSpread({}, parentNavigationHelpers, {}, helpers, {
      dispatch,
      emit: emitter.emit,
      isFocused: parentNavigationHelpers ? parentNavigationHelpers.isFocused : () => true,
      canGoBack: () => {
        const state = getState();
        return router.getStateForAction(state, CommonActions.goBack(), {
          routeNames: state.routeNames,
          routeParamList: {}
        }) !== null || (parentNavigationHelpers === null || parentNavigationHelpers === void 0 ? void 0 : parentNavigationHelpers.canGoBack()) || false;
      }
    });
  }, [router, getState, parentNavigationHelpers, emitter.emit, onAction]);
}
